classes/auto_login.php

```diff
@@ -0,0 +1,212 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Auto-login class.
+ *
+ * @package   auth_saml2
+ * @copyright 2020 The Open University
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace auth_saml2;
+
+use auth_saml2\admin\saml2_settings;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Auto-login class.
+ */
+class auto_login {
+    /** @var bool True if already considered auto-login on this page. */
+    protected static $considered;
+
+    /**
+     * Called on every page to give an opportunity to auto-login.
+     *
+     * This is called on most pages via the after_require_login callback, and also by the
+     * before_http_headers callback, so it might be called twice. We only want to run it once.
+     */
+    public static function consider() {
+        global $SESSION, $SCRIPT;
+
+        if (self::$considered) {
```

Comment:
this method got quite a few if statements and I can't find test cases for each of them. Would be good to cover those in bahat or php unit. Like emulate POST request (when ready for autologin) and check that we actually don't auto login and etc

Reply:
Good point. I've added a new Behat scenario that tests most of the possible reasons including POST request, if the plugin is disabled, if the option is turned off.